### DISCUSSION TITLE
Discord server?

### DISCUSSION THREAD

**Discussion by nnmrts** [unifiedjs/unified#127] [General]
I'm personally and generally against creating a discord server for every little group or project, but the unified ecosystem is huge, unique to itself and spans a lot of different areas of the javascript world, and there (unless I missed something) doesn't seem to be a main place for the community to have a fast-paced and light-hearted chatting experience about everything related to unified. Sure, there are the `awesome-` lists and this discussion forum for the community to chat and show off their projects, but I feel like these don't offer the same freedom and diversity of discussion types you would have on something like a discord server.

What do you think about this?

**Comment by wooorm** [comment]
We used to have a couple Gitter places, a long time ago. Then we got Spectrum, which is more about discussions, which since moved to GH Discussions. What would the benefit be of having a Slack/Discord like thing be? And if we’re going for it, (why) should it be Discord?

**Comment by ChristianMurphy** [comment]
The community has tried a few platforms https://gitter.im/remarkjs and https://spectrum.chat/unified are two from recent times, there may be more from further back.

There is a tradeoff with a chat based approach vs a forum style approach (like GitHub discussions).
Chat is really good a informal quick conversations, however it tends to be not-so-great for questions (threads are more difficult to get right, and linking repeat questions, and marking questions as answered is also difficult).
Forum style chat tends to be better for questions, it has clear threads, and often comes with a voting mechanism, questions can be clearly marked as answered, and are easy to search/link for future reference.

Over the past few years, most community conversations seems to be questions, at about a ratio of 3-4 questions to 1 of any other type of conversation.
I'm interested in fostering more types of conversation, and a chat platform may be a piece of that.
I'd be interested to hear thoughts around how questions could be effectively steered back to the forums, without derailing conversations.

**Reply by wooorm** [reply]
good points!

**Comment by nnmrts** [comment]
The goal shouldn't be and the result definitely won't be a full replacement of current ways to chat and communicate. Discord is just great for not-so-important stuff but can also sometimes help with a bit more complicated but "fast-paced" stuff. Things like big pull requests sometimes require a series of back-and-forths about small things like typos, which (at least for me) feel wrong in the form of several issue or pr comments on GitHub. 

I don't think most people would misunderstand a Discord server as an alternative to GitHub issues. And even if they do, the maintainers (and other people) can always judge on a case-by-case basis if something is a quick question or a full-blown bug report. I also don't think a lot of people even expect to get the types of responses they would get on GitHub. And if they do, just say: "Can you open an issue for that?"

Also, I don't know how new it is, but Discord has a "reply"-system too now, which makes it a bit better in terms of following a discussion while the conversation itself may derail.

I personally think a Discord server would bring several benefits with it:
- "user-driven" support
    - If I read a particular question in a discord server and have time and the know-how at that moment, I'll probably try to help that person out even though I'm not the maintainer. I wouldn't do that on GitHub issues or discussions, especially because I don't even open most posts there as a non-maintainer. This would also take that time off maintainers, who would have to answer these questions in issues or discussions instead.
- A place for the community to show off stuff to ask for feedback
    - Can bring the community closer together and definitely helps the growth of unified.
- A way to clarify small things way faster.
    - Which also means you don't have to write everything in a precise manner.
- A lot of people already have it, and I think even if you don't have an account, you can interact with a server in limited ways.
- It is free and works great everywhere.

Starting a Discord server wouldn't be "a chat based approach vs a forum style approach", but a chat based approach AND a forum style approach instead of just a forum style approach. I don't think one or the other would die off, because the other exists, I think most people can distinguish between the two and choose the right one if they have a question or found an issue.

However I would always make sure to describe the different ways whenever a readme or wiki or website links to them. Like:

Found a bug or thought about a useful feature? Go to [Issues].
Want to know more or talk about specifics, certain design decisions or the future of the project? Go to [Discussions].
Wanna hang out with the community and chat about everything related (or not so related) to unified? Go to [Discord].

**Reply by ChristianMurphy** [reply]
> "user-driven" support
> If I read a particular question in a discord server and have time and the know-how at that moment, I'll probably try to help that person out even though I'm not the maintainer. I wouldn't do that on GitHub issues or discussions, especially because I don't even open most posts there as a non-maintainer. This would also take that time off maintainers, who would have to answer these questions in issues or discussions instead.

I appreciate the sentiment, I haven't seen this ideal translate to reality.
Here's an example of a chat channel in the React community, with 32K developers able to respond https://spectrum.chat/react?tab=chat
It's a stream of questions, with no answers, compare that with threads (the interface which has now become GitHub discussions) https://spectrum.chat/react/help?tab=posts, where most questions are answered. Even there with a community of 32K most of the answers end up coming from about 10 people :sweat_smile: 

> place for the community to show off stuff to ask for feedback
>    Can bring the community closer together and definitely helps the growth of unified.

GitHub discussions has the " :bulb: idea" and " :pray: show and tell" tags for this as well.

> A way to clarify small things way faster.
> Which also means you don't have to write everything in a precise manner.

A nice ideal, but again doesn't translate as well in practice.
Across all Q&A and Support communities the [XY Problem](https://meta.stackexchange.com/questions/66377/what-is-the-xy-problem) is prevalent.
A nice part of GitHub issues, and hopefully soon GitHub discussions as well, is the ability to have templates that guide people towards the information needed to actually answer a question.

Without that context/precision, maintainers end up asking the same questions "what package(s) are you using?", "what version(s) are you using?", "what version of node or browser are you using?", and "what is the input and the desired output?" to every question that gets asked.
It's a process which is excruciatingly time consuming and brutally repetitive, to the point of nearly causing maintainer burn out.

> I think most people can distinguish between the two and choose the right one if they have a question or found an issue.

Again, I appreciate the optimism, I'm not sure I share it.
Taking https://github.com/remarkjs/react-markdown/issues?q=is%3Aissue+is%3Aclosed as an example, even with guidance as to where to direct questions and how to search for previous answers embedded in the template and linked in the community guidelines.
Over half of the issues, from otherwise well meaning and intelligent people, ignore the guidelines, and open duplicate issues or ask questions in the issue tracker instead of the forum.

---

What a lot of this boils down to is a combination of a few things:
* abstract syntax trees require building some domain knowledge to work with, the guides at https://unifiedjs.com/learn/ are working to help build this out some, additional guides and more ideas are welcome
* The Unified community mostly communicates in an asynchronous manner, past attempts to be more synchronous have not been successful, there may be value in considering synchronous communication again, but a tool (Discord, Slack, etc) alone likely isn't enough to make it sustainable.
* Community and maintainer time is finite and limited, things have to be prioritized to be respectful of both. Chat's inherent value comes from being synchronous. Without explicit structure, it comes with an implicit [24/7/365](https://www.macmillandictionary.com/buzzword/entries/24-7-365.html) availability expectation, which is neither accurate nor beneficial.

Which is why I'm cautious, and lean somewhat against, reintroducing chat, whether through Discord or another service.

---

Taking a step back, to see if there is an effective way forward.
I'm going to try to summarize your overarching points (definitely jump in if I miss something or misrepresent it)

1. There may be value in having a synchronous channel for communication
2. There is an interest in having new avenues/paths towards maintainership
3. Issues/Discussions can be perceived as being formal or intimidating

---

To the first, I'm open to the idea, but would want clearer expectations on when people are available and for how long.
Perhaps something like https://simonwillison.net/2021/Feb/19/office-hours

For the second point, this is intended to be addressed by:

> Typical participation of an organization team maintainer includes:
>
> * Answering questions and helping novice contributors
> * Contributing code, non-code, and documentation changes that improve the scope of the team
> * Reviewing and merging pull requests
> * Participating in initiatives and discussions
>
> Maintainers are not required to contribute code. It is, however, required that maintainers have actively participated for a consistent and significant amount of time already.

https://github.com/unifiedjs/collective/blob/HEAD/members.md#participation

If there is a perceived gap, I'd be interested to hear thoughts on what that is and how to address it.

To the third point. I'm open to suggestions on how to make both discussions and issues more approachable.
I know @wooorm was looking into the possibility of a bot to help guide new contributors through the open source process as one approach.
There are discussions around how to trim back the issue and feature templates where possible.
I've been adding code sandboxes where possible as an easy starting point for replicating issues and sharing features.
There may be other gaps or sore points, insights are welcome!

**Reply by nnmrts** [reply]
I definitely understand your perspective, but I don't think starting a Discord server would be that dramatic. In it's simplest sense, it would just be a "hub", a more concentrated and organized way of "talking about everything related to it" than using the hashtag #unified on stackoverflow or twitter. There doesn't need to be someone in there answering questions or even feeling obliged to do so.

I actually don't think it would be **that** big of a deal at the end. It's free, accessible and widely accepted as yet another way to talk about a project and announce stuff to your community, the latter can even be automated thorugh GitHub releases or projects or milestones.

**Reply by ChristianMurphy** [reply]
> It's free

The opportunity cost is not zero, the server is monetarily free, but moderation and curation in a community takes time, that time has implicit cost, which grows for each new platform that needs curation and moderation.

> I don't think starting a Discord server would be that dramatic
> I actually don't think it would be that big of a deal at the end.

Unified is a [stadium project](https://645ventures.com/github-at-scale-and-how-to-help-stadium-model-maintainers), with somewhere around 389,796 adopters today [(lower bound 771; upper bound 5,463,604+ on the `unified` package alone)](https://www.npmjs.com/package/unified), with [8 maintainers](https://github.com/unifiedjs/collective#unified-team).
That's a ratio of 48,724 adopters to 1 maintainer.
So no, I don't think creating more work for the team is "not that big a deal" or being "dramatic", it requires consideration into the costs versus the benefits.

> There doesn't need to be someone in there answering questions or even feeling obliged to do so.

I'm not sure what you're expecting here.
As I shared above: this is what a channel that doesn't get curated or answers looks like: https://spectrum.chat/react?tab=chat
From my perspective that is not a "fast-paced and light-hearted chatting experience", "A way to clarify small things way faster", or "a more concentrated and organized way of talking about everything related to it".

> In it's simplest sense, it would just be a "hub", a more concentrated and organized way of "talking about everything related to it" than using the hashtag #unified on stackoverflow or twitter.

Doesn't Github Discussions also serve the purpose?

**Reply by nnmrts** [reply]
> but moderation and curation in a community takes time

Yeah, I know that and chose my words carefully. I moderated communities like the german part of Genius for years, and that is an actual forum like we're in 2007.

Discord isn't such a platform though. Moderation can be automated and curation is only a one time thing, which works forever if well thought out. The comparison to React's Spectrum chat is weak. Also it's not even referenced on the React homepage, seems like they realized it isn't what they thought it would be (or maybe it's just here because Spectrum is built on React). Sure, if we start a Discord server with only 4 unorganized channels and no "#welcome" or "#rules" channels whatsoever, then the outcome will be the same. The experience for new users is completely different. When I click on spectrum.chat/react?tab=chat, the first thing I read is some random question. Also Spectrum itself is widely different to Discord in a lot of aspects as well.

When you click on an invite link of a good Discord server, you first get onto a welcome screen that probably requires you to read the rules before even getting access to all the channels. I'm sorry, but at this point I don't really get the direction of the discussion here anymore. It's not like we're launching a product, it's really just a Discord server. And I'm especially surprised because we're all coming from an open-source perspective here. All you said is true for every new unified repo as well.

If people come to a discord server of an open source project and expect enterprise level support, it's literally their fault. You can't do more than to formulate a description of the purpose of a platform, no matter what platform that is. And the moderation and curation can be open- and repos **and** a Discord server, which is fully understandable.

Please, do me a favour and look at the (somewhat community-run) [Reactiflux Discord](https://discord.com/invite/reactiflux), a platform the React homepage actually links to. It has 100k+ users, of which 8k are regularly online. What do you think about it?

I don't see how a unified Discord server on a smaller scale would be different. I bet a lot of people in the community (including me) are ready and motivated to start such a server. If I'm totally honest, I didn't expect this discussion to get replies arguing against opening a Discord server in general and really I wonder if you would think about all that the same once you look into Reactiflux for example, or any other well thought out Discord server of a well thought out open-source project.

**Reply by remcohaszing** [reply]
I agree with @nnmrts that Discord works very well for quick back-and-forth conversations. I don’t think the comparison to Spectrum is a fair one. In my experience Spectrum _feels_ like a forum, whereas Discord _feels_ like a chat. The bar for asking a question is much lower. It’s more similar to Slack, but with more advanced IMO.

I use Discord for chat, voice calls, and screen sharing on a daily basis. I’ve found it to be much more useful than alternative chat services.

There are some examples of such communities where users actually help each other, i.e. TypeScript Discord, Babel Slack, and Kubernetes Slack. The downside is this attracts a lot of people who simply don’t read the documentation, but turn to the chat instead. This means the same questions will be asked over and over.

I agree with @ChristianMurphy that such a server requires maintenance because of moderation. I don’t think the unified team should be burdened with this responsibility if they don’t want to.

I think having an official unified Discord server is fine, as long as someone wants to maintain it (@nnmrts perhaps?) and it’s clear people probably won’t get answers from official unified maintainers.

**Comment by wooorm** [comment]
I see things less rosey as @nnmrts, and less bleak as @ChristianMurphy. I think @remcohaszing sums it up nicely.

The amount of folks not reading release notes, or readmes, or knowing that they can use CSS to style their React, is giant. Lowering the bar by adding chat, which promises a quick and direct answer, without having to fill out an issue template, *will* result in a ton of already answered/documented questions.
I do know that I personally, and knowing @ChristianMurphy he will too, even if we don’t *have* to respond immediately, we will do our utmost best to respond on weekends and in our evenings.
@remcohaszing’s suggestion of having it “unofficial” might help force this.

I do agree that Discord is not Spectrum.

One point I have not seen made before, is that I’m guessing that GH is probably working on a discord fork to integrate into their platform. Transitioning in a year or so might be unfortunate.

For the comparison to reactiflux: well, that’s a Discord specifically about helping folks. Could we discuss what new aspects (so, other than helping folks out) would a unified Discord provide?
